### PR TITLE
API-breaking change: Deprecate ConnectionProvider with new RuntimeProvider

### PR DESCRIPTION
**Pull request by XOR-op**
# Why we need new interface?
Current `ConnectionProvider` interface didn't provide enough ability for end-users to create customized connections, although it looked like. For example, users find it hard to customize underlying TCP or UDP sockets, see #1863, #1870, #1335. Users may also seek for abstract TCP stream instead of real one. Related discussion can be found on #1863.

# What is our plan?
We design new `RuntimeProvider` interface, whose responsibility is to provide TCP-like and UDP-like connections. Currently, the new interface looks like: 
```rust
pub trait RuntimeProvider: Clone + Send + Sync + Unpin + 'static {
    type Handle: Clone + Send + Spawn + Sync + Unpin;
    type Timer: Time + Send + Unpin;
    type Udp: DnsUdpSocket + Send;
    type Tcp: DnsTcpStream;

    /// Create a runtime handle
    fn create_handle(&self) -> Self::Handle;

    /// Create a TCP connection with custom configuration.
    fn connect_tcp(
        &self, server_addr: SocketAddr,
    ) -> Pin<Box<dyn Send + Future<Output = io::Result<Self::Tcp>>>>;

    /// Create a UDP socket bound to `local_addr`. The returned value should **not** be connected to `server_addr`.
    /// *Notice: the future should be ready once returned at best effort. Otherwise UDP DNS may need much more retries.*
    fn bind_udp(
        &self, local_addr: SocketAddr, server_addr: SocketAddr,
    ) -> Pin<Box<dyn Send + Future<Output = io::Result<Self::Udp>>>>;
}
```
where we introduce new `DnsUdpSocket` as the counterpart of `DnsTcpSocket` ,removing requirements for bind and connect.

# Design problems
- <del>Should we take `GenericConnection` as our only implementation, or leaving some interface for users to implement their own ones? </del>
# Progress
- Underlying support
  - [x] support for providing customized TCP stream
  - [x] support for providing customized UDP socket for quic
  - [x] support for providing customized UDP socket for udp
- [x] change interface
- [x] fix dependent code
- [x] fix tests
- [x] pass tests

### CODE REVIEW

**Comment by XOR-op** [comment]
@bluejekyll @djc Do you think it's ok to make `GenericConnection` the only implementation of new interface? This would significantly reduce the complexity of code, and the original flexibility will be offered by new `RuntimeProvider` interface.

**Comment by djc** [comment]
@XOR-op I need a bit more context on that question. Can you link to the relevant locations in the code and explain both alternatives in slightly more detail?

**Comment by XOR-op** [comment]
The `GenericConnection` is in
https://github.com/bluejekyll/trust-dns/blob/3235d2b3f6433271f71780963936edab541c2ff3/crates/resolver/src/name_server/connection_provider.rs#L352-L363

which is used in the default implementation of `GenericConnectionProvider`:
https://github.com/bluejekyll/trust-dns/blob/3235d2b3f6433271f71780963936edab541c2ff3/crates/resolver/src/name_server/connection_provider.rs#L97-L99
https://github.com/bluejekyll/trust-dns/blob/3235d2b3f6433271f71780963936edab541c2ff3/crates/resolver/src/name_server/connection_provider.rs#L108-L115
Since @bluejekyll wants to get rid of `ConnectionProvider` interface, I expect he does not think this layer of abstraction is necessary if we have new interface (Correct me if I misunderstand). Therefore, as `GenericConnection` has been the defacto default implementation, we can just use it. Otherwise, we may need to design another interface for it, and I am not sure whether it's worth.

**Comment by bluejekyll** [comment]
Yeah, I think this makes sense. The only thing that really leverages this generic are some tests, I believe. All the original concepts for it have really leverages the Runtime interface. So yes, I like this idea, but I do think you may end up with having to do a lot of work to fix tests that use it today.

**Comment by XOR-op** [comment]
> Yeah, I think this makes sense. The only thing that really leverages this generic are some tests, I believe. All the original concepts for it have really leverages the Runtime interface. So yes, I like this idea, but I do think you may end up with having to do a lot of work to fix tests that use it today.

@bluejekyll I have an idea, that we can introduce a test-only feature, maybe called `mocktest` or something. Then we can conditionally compile related code, just like using macro to compile in C/C++. But I'm not sure whether you will like this idea. If you think it is ok, I could give it a try.

**Comment by XOR-op** [comment]
Hi @bluejekyll @djc , this is the final version of `RuntimeProvider` . Now it looks like

```rust
pub trait RuntimeProvider: Clone + Send + Sync + Unpin + 'static {
    type Handle: Clone + Send + Spawn + Sync + Unpin;
    type Timer: Time + Send + Unpin;
    type Udp: DnsUdpSocket + Send;
    type Tcp: DnsTcpStream;

    /// Create a runtime handle
    fn create_handle(&self) -> Self::Handle;

    /// Create a TCP connection with custom configuration.
    fn connect_tcp(
        &self, server_addr: SocketAddr,
    ) -> Pin<Box<dyn Send + Future<Output = io::Result<Self::Tcp>>>>;

    /// Create a UDP socket bound to `local_addr`. The returned value should **not** be connected to `server_addr`.
    /// *Notice: the future should be ready once returned at best effort. Otherwise UDP DNS may need much more retries.*
    fn bind_udp(
        &self, local_addr: SocketAddr, server_addr: SocketAddr,
    ) -> Pin<Box<dyn Send + Future<Output = io::Result<Self::Udp>>>>;
}

```

This version has a little difference with the first draft: I replace arguments from `NameServerConfig` and `ResolverOpt` to `addr`s, for two reasons: 

1. `server_addr` plus with `local_addr` is enough for connection creation. In my opinion, `bind_addr`for client is useless, and it doesn’t work as expected on Linux (whose original goal is for customized routing). It’s recommended to use functions like `bind_to_device` to achieve this goal, which can be easily implemented with this interface.
2. Because of how UDP socket is implemented, it’s too difficult to pass configs to underlying implementation.

For other code changes, I did the following things:

- add `_with_future` constructors to various connections
- QUIC connections now are provided by `RuntimeProvider`
- Random UDP ports are guaranteed, even if customized runtime provider is offered (unless provider ignores input explicitly)
- fix tests

What I didn’t change:

- Although I want to get rid of `GenericConnection`, I failed. It exists for compatibility with previous tests.

Hope to see your review and feedback!

**Comment by bluejekyll** [comment]
Sorry, I missed that this was completed. I'm reviewing, will update.

File: crates/async-std-resolver/src/net.rs
Comment (djc) [inline crates/async-std-resolver/src/net.rs:81]: I don't think we want to just leave this here?

File: crates/resolver/src/name_server/name_server.rs
Comment (djc) [inline crates/resolver/src/name_server/name_server.rs:46]: I'm not sure this makes sense. Why do we have an `AbstractNameServer` and a `NameServer`? If anything, I feel like the naming here should be reversed, more like `NameServer` and `GenericNameServer` (or something like that).

File: crates/resolver/src/name_server/name_server.rs
Comment (djc) [inline crates/resolver/src/name_server/name_server.rs:33]: `RuntimeProvider` already requires `Send`, so this trait bound is no longer needed?

File: crates/resolver/src/name_server/name_server.rs
Comment (djc) [inline crates/resolver/src/name_server/name_server.rs:215]: In general, please don't do stylistic changes (like moving bounds from the `impl` to a `where` clause) at the same time as semantic changes, this makes it harder to review the code.

File: crates/resolver/src/name_server/name_server.rs
Comment (djc) [inline crates/resolver/src/name_server/name_server.rs:275]: Why is this necessary. If it's just for mocking, can it be scoped to a `#[cfg(test)]` block?

File: crates/resolver/src/name_server/name_server_pool.rs
Comment (djc) [inline crates/resolver/src/name_server/name_server_pool.rs:34]: Here, too, I think we should switch the naming around, to leave this called `NameServerPool`. At least, you mentioned `GenericConnection` was mainly used for testing, right?

File: crates/resolver/src/tls/dns_over_rustls.rs
Comment (djc) [inline crates/resolver/src/tls/dns_over_rustls.rs:78]: Why do we need this extra layer of abstraction?

File: crates/proto/src/udp/udp_stream.rs
Comment (djc) [inline crates/proto/src/udp/udp_stream.rs:71]: Why do we have a separate `UdpSocket` and `DnsUdpSocket` traits?

File: crates/proto/src/udp/udp_stream.rs
Comment (djc) [inline crates/proto/src/udp/udp_stream.rs:32]: This seems pretty ugly, why do we need this?

**Review by djc — COMMENTED** [review]
I think the high-level trait definition looks pretty good, but this PR seems to come with a bunch of extra abstractions for which I don't quite understand why they're needed.

File: crates/proto/src/udp/udp_stream.rs
Comment (XOR-op) [reply crates/proto/src/udp/udp_stream.rs:71]: `UdpSocket` must implement `bind()`. However, since we use `RuntimeProvider` to create such a "connection", it only requires `send_to()` and similar methods. In other words, `DnsUdpSocket` only needs to transfer data, without caring how to establish the "connection".

File: crates/proto/src/udp/udp_stream.rs
Comment (XOR-op) [reply crates/proto/src/udp/udp_stream.rs:32]: It's surely ugly, but without it I don't know how to comply to current UDP DNS behavior. For UDP DNS, we connect through a new UDP socket for each request. Use this closure. Everytime the `NextRandomUdpSocket` requires, it invokes the closure. Another approach I came up with just now is moving the `RuntimeProvider` interface to `proto` crate (since udp related codes are there). What about your opinion?

File: crates/resolver/src/name_server/name_server.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server.rs:46]: Since the original idea of `NameServer` and `GenericNameServer` comes from the need of mock testing. We expose it since the tests need it, not because we expect end users to actually use it. Rename it could make it clear in documentations.

File: crates/async-std-resolver/src/net.rs
Comment (XOR-op) [reply crates/async-std-resolver/src/net.rs:81]: This is a TODO from the original code. It has little connection with my PR and I don't know that problem much. So I just leave the TODO here.

File: crates/resolver/src/name_server/name_server.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server.rs:215]: Sorry for that. I just thought this change will make code clear and won't cause some confusion.

File: crates/resolver/src/name_server/name_server.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server.rs:275]: Because I introduce this trait as an intermediate layer. If we didn't need testing, we could remove this redundant trait and move the code to actual implementation. Therefore, it can not be labeled as `test`. I will modify this confusing comment.

File: crates/resolver/src/name_server/name_server_pool.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server_pool.rs:34]: Not accurate. The `GenericConnection` is the actual implementation. My previous question is actually "Can I remove the `GenericConnection` by moving its code to proper places". But I cannot get rid of it at last.

File: crates/resolver/src/name_server/name_server.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server.rs:33]: You are right. I will resolve it.

File: crates/resolver/src/tls/dns_over_rustls.rs
Comment (XOR-op) [reply crates/resolver/src/tls/dns_over_rustls.rs:78]: I didn't get what you mean. Could you make some explanation?

File: crates/async-std-resolver/src/net.rs
Comment (djc) [reply crates/async-std-resolver/src/net.rs:81]: Ah, that's fair.

File: crates/resolver/src/name_server/name_server.rs
Comment (djc) [reply crates/resolver/src/name_server/name_server.rs:275]: Why do we need this intermediate layer, though?

File: crates/proto/src/udp/udp_stream.rs
Comment (djc) [reply crates/proto/src/udp/udp_stream.rs:32]: Yeah, IMO it would be better to move `RuntimeProvider` into the proto crate.

File: crates/proto/src/udp/udp_stream.rs
Comment (djc) [reply crates/proto/src/udp/udp_stream.rs:71]: So would this also be solved by moving `RuntimeProvider` into proto?

File: crates/resolver/src/tls/dns_over_rustls.rs
Comment (djc) [reply crates/resolver/src/tls/dns_over_rustls.rs:78]: It's not clear to me why we need this extra function.

Comment (XOR-op) [reply crates/resolver/src/tls/dns_over_rustls.rs:78]: This is the replacement of `new_tls_stream`. I think I should remove the former one since it's not used at all. If you want to know why the original function is needed, it's because we have rustls, native-tls and openssl implementations.

File: crates/proto/src/udp/udp_stream.rs
Comment (XOR-op) [reply crates/proto/src/udp/udp_stream.rs:32]: OK, I will work for that.

File: crates/proto/src/udp/udp_stream.rs
Comment (XOR-op) [reply crates/proto/src/udp/udp_stream.rs:71]: Nope. And this should make sense naturally, because we have `Connect` and `DnsTcpStream` counterpart for TCP in original code.

File: crates/resolver/src/tls/dns_over_rustls.rs
Comment (djc) [reply crates/resolver/src/tls/dns_over_rustls.rs:59]: Right, if we can remove `new_tls_stream()` because it's gone unused we should definitely do that.

File: crates/proto/src/udp/udp_stream.rs
Comment (XOR-op) [reply crates/proto/src/udp/udp_stream.rs:32]: Hi @djc, I tried to fix the probelm with `RuntimeProvider` as struct member. However, to remove this we need to modify `UdpStream` and `UdpClientStream` much. Since `UdpStream` and `UdpClientStream` is widely used among the project, such an introduction will cause lots of break in other places, and probably other ugly codes. I think although this definition looks scary, it is in fact a simple closure, with two `SocketAddr` as arguments and one `Future` as result. The other parts just make compiler happy. It is also private within crate. After the trial my intention is to keep the "ugly" definition. But I also want to hear about your suggestion.

File: crates/resolver/src/name_server/name_server.rs
Comment (XOR-op) [reply crates/resolver/src/name_server/name_server.rs:46]: However, if you insist the original naming is better, I could revert it back.

**Comment by XOR-op** [comment]
@bluejekyll Hi, it has been more than a week after your reply. Do you have any updates on the review? I'd like to hear your feedback.

**Comment by hottea773** [comment]
@XOR-op / @bluejekyll / @djc any updates on this getting merged?

(And thanks for this @XOR-op I've been using it and it's fixed my issue #1870 excellently!)

**Comment by XOR-op** [comment]
> @XOR-op / @bluejekyll / @djc any updates on this getting merged?
> 
> (And thanks for this @XOR-op I've been using it and it's fixed my issue #1870 excellently!)

Glad to hear that! There are still 2 remaining issues to discuss: the first is whether the renaming of `GenericNameServer` etc is appropriate; the second is whether we should make `UdpCreator` looks better. I'm still waiting for comments from @djc and @bluejekyll . Without their approvals, this PR cannot be merged.

**Comment by djc** [comment]
Probably good to proactively get this rebased.

**Review by bluejekyll — COMMENTED** [review]
I'm working through the final review... about halfway. Everything looks good so far.

File: crates/proto/src/native_tls/tls_client_stream.rs
Comment (XOR-op) [reply crates/proto/src/native_tls/tls_client_stream.rs:73]: Sorry I don't quite get what you mean. Do you mean that this interface lack `client_config: Arc<ClientConfig>`? It's because this is native_tls and `ClientConfig` is for rustls only.

**Review by bluejekyll — APPROVED** [review]
Ok, I think this cleans things up a bit. I think we can do more, but this changelist is already quite big.

File: crates/proto/src/native_tls/tls_client_stream.rs
Comment (bluejekyll) [reply crates/proto/src/native_tls/tls_client_stream.rs:73]: Really just the name.

Comment (bluejekyll) [reply crates/proto/src/native_tls/tls_client_stream.rs:73]: It's fine.

**Comment by XOR-op** [comment]
@bluejekyll It looks like we need to bump the MSRV in the next release.

**Comment by bluejekyll** [comment]
Yeah, though 1.65 is pretty recent... it's just 2 releases ago. 1.68 will stabilize in a week. @djc, any concerns?

**Comment by djc** [comment]
Why do we need the MSRV bump? I would not push my crates to 1.65 yet.

**Comment by XOR-op** [comment]
> Why do we need the MSRV bump? I would not push my crates to 1.65 yet.

In my opinion, compared to the bumped MSRV, an API-breaking change would have more influences. And lots of people would adopt 1.65 because GAT was stablized in this version. Besides that, I saw lots of MSRV bumps in our changelog, so I just expect that a MSRV bump is natural for a new release.

**Comment by XOR-op** [comment]
@djc @bluejekyll GAT has been removed. Now there is no need for bumping MSRV.

**Comment by XOR-op** [comment]
@djc @bluejekyll Please approve the CI workflow. Thanks.

**Comment by XOR-op** [comment]
@bluejekyll @djc Now all the requirements have been met. It’s time to merge it. Thanks for all your help with the PR.

**Review by bluejekyll — APPROVED** [review]

**Comment by bluejekyll** [comment]
@XOR-op thank you for taking the time and energy for getting these changes in. I know they were large. @djc, with the 1.65 requirement gone, I think this is ready to land. I'm going to merge now.

**Comment by jeff-hiner** [comment]
I've been trying to port over some code that used ConnectionProvider to expose an authenticated DoH functionality. The existing code against 0.22 implements a custom ConnectionProvider to do the TLS handshake and ensure the DoH requests have the appropriate auth header. We then create an AsyncResolver using this connection provider to do specific lookups against a private resolver.

Looking at how drastically this PR has changed everything, it's completely unclear how I'm supposed to make that functionality work under your model. This refactor appears to completely break my use case. Can you give some documentation on how with this new model I'm supposed to override how connections are created?

**Comment by bluejekyll** [comment]
@jeff-hiner thanks for your feedback. This was a concern of mine. The intention was to move all of the underlying socket construction up to the runtime provider to simplify porting between runtimes.

Would it be possible to share a link to your implementation so that we can take a look at understanding what we got wrong here? @XOR-op possibly an area that you can help give some guidance?

**Comment by mokeyish** [comment]
I have implement socks5 proxy over the new RuntimeProvider.You can refer here to see🤣.

https://github.com/mokeyish/smartdns-rs/blob/1d3b7c5bda55c887add46467cd046a272ea3d009/src/dns_client.rs#L928

**Comment by XOR-op** [comment]
@jeff-hiner Hi Jeff, it could be hard to use `RuntimeProvider` interface to implement your need. However, I think the `AbstractNameServer` in `resolver/src/name_server/name_server.rs` could satisfy your need. It resembles the original `ConnectionProvider` interface and should be easy to port.

**Comment by jeff-hiner** [comment]
I'll take a look at it and see if I can get it working.

@bluejekyll Unfortunately the code I'm trying to port is proprietary so I can't directly post a link, but if I can't figure out how to make it work with `AbstractNameServer` I'll work on getting an abbreviated version posted here.

**Comment by bluejekyll** [comment]
@jeff-hiner, If this continues to be an issue for you, maybe there are some ways for us to rework the new system to make this easier?

At the least, your explorations will be helpful in documenting for others what needs to happen in order to update to the new interfaces.

**Comment by jeff-hiner** [comment]
Thanks for being willing to help out! I've been swamped with other work this week and last, but this is absolutely on my to-do list. Hopefully it'll be straightforward.

**Comment by XOR-op** [comment]
@jeff-hiner Is there any update? Now the docs about `AbstractNameServer` says it's used only for mocking. But if people really use them, I need to clarify it to avoid confusion.

**Comment by jeff-hiner** [comment]
Yeah the original docs for that had remarks about it being used mostly for mocks and testing, but in reality it's a critical component for being able to tweak DoH connection factories.

Speaking of factories, I'm running into problems with the requirement for `CreateConnection` and `DnsHandle` to be impl on the same type. In my mental model `new_connection` is a factory that returns new `DnsHandle`s.

The problem is the factory contains some metadata (in this place a closure that produces an https token) but actually creating a connection requires the `new_connection` args. Specifically I need the TLS name and socket addr to make a `DnsHandle` which means it cannot be the same factory type as `CreateConnection`. I can contain one type within the other, but I can't return an identical type with the info populated.

```
fn new_connection(
    &self,
    config: &NameServerConfig,
    _options: &ResolverOpts,
) -> Result<DohHandle<T>, ResolveError> {
    let hostname = config
        .tls_dns_name
        .as_ref()
        .ok_or("DoH doesn't work without a hostname")?;
    let addr = config.socket_addr;

    let client = reqwest::ClientBuilder::new()
        .resolve(hostname, addr)
        .connect_timeout(Duration::from_secs(10))
        .build()?;
    let mut host_url = reqwest::Url::parse(&format!("https://{}", hostname))?;
    host_url.set_path("dns-query");

    Ok(DohHandle {
        client,
        host_url,
        howto_token: self.howto_token.clone(),
    })
}
```

**Comment by djc** [comment]
@jeff-hiner thanks for taking the time to look into it! Would you be able to submit a PR with a refinement of the new API towards something that would address your project's needs?

**Comment by jeff-hiner** [comment]
In 0.22 the P parameter encapsulates `DnsHandle` already:

```
AsyncResolver<C, P> where:
C: DnsHandle<Error = ResolveError>,
P: ConnectionProvider<Conn = C>,
```

With the new model it's this:
```
AbstractNameServer<C, P> where:
C: DnsHandle<Error = ResolveError> + Send + Sync + 'static + CreateConnection,
P: RuntimeProvider
```

`DnsHandle` in both cases is what lets you `send` requests and get responses. There are places in the code now where `impl<C, P> DnsHandle for AbstractNameServer<C, P>` which is just utterly confusing to me. `AbstractNameServer` is not a `DnsHandle`, it creates them.

I think one problem is there are assumptions that a DNS "connection creator" is going to have to do raw TCP and UDP bind/connect. That's not true when implementing a DoH resolver for my own use because I'm not writing `hyper` or `reqwest` from scratch-- I'm deliberately making assumptions about the runtime I'm using, and I use higher level abstractions for https. The runtime provider itself isn't directly relevant to me, nor does it have any bearing on the handle or connection creator. So already the model falls apart there.

In all honesty, when I'm implementing `AbstractNameServer` for a real program I don't care about supporting multiple runtimes anyway.

What I want to see is to change the types slightly. Something like this:

```
AbstractNameServer<H, C> where:
H: DnsHandle<Error = ResolveError> + Send + Sync + 'static,
C: CreateConnection<H>,

fn new_connection<C>(
        handle_factory: &C,
        config: &NameServerConfig,
        options: &ResolverOpts,
    ) -> Box<dyn Future<Output = Result<H, ResolveError>> + Send + Unpin + 'static>;
```
and provide a stock
```
impl CreateConnection<H> for RuntimeProvider
```
that does normal UDP+TCP DNS using `connect_tcp` and `bind_udp` and exposes a way to create "DNS handles". Then you don't need to add dynamic downcasting or unwraps for types that don't line up how we expect.

**Comment by XOR-op** [comment]
@jeff-hiner For your first question, it's not my change. You could find `NameServer` implements `DnsHandle` in 0.22. If you have opinions about this, you could open another issue to discuss it.

For your second question with TCP and UDP, I think that's because of the difference of how we use trust-dns. Many users (including me and authors of all referred issues) just simply use trust-dns. We just want a DNS solution, and we prepare all the network-stack stuff for it. Your case is different since you operate inside trust-dns. The runtime provider itself isn't directly relevant to you but relevant to us. So it's over-designed for your need but appropriate for our need. Also, you could easily rely on the infrastructure trust-dns offers by embedding a `TokioRuntimeProvider`. I don't think you need to re-implement them yourself.

And for your last draft, I don't see any difference with the original design in 0.22. It's the inconvenience the old design brings that motivates me to open this PR. `GenericConnection` takes care of all UDP, TCP, DoH, DoT and DoQ things, and few of developers are willing to re-implement them just to customize something. In your case you also only implement DoH.

In honesty, it's very difficult to provide some ways of customization in trust-dns. To change one API I must fix dozens of files. What I could only do is to balance the new API design and the legacy code. That's why some APIs may look complicated or confusing.

**Comment by jeff-hiner** [comment]
Maybe it's just confusion from naming, then. I don't know if @djc has some context on why it's designed this way, maybe that would help me understand.

I see that there's a desire to remain compatible across both tokio and async-std, and that's where a lot of the complexity and pain comes from. There's a scratch-built http/2 client in `https_client_stream.rs` where most folks would just use `hyper`, because of course `hyper` depends on `tokio`. That part makes sense. It's rough, by necessity.

But because it's all bespoke, I can't just tweak `GenericConnection` to inject a header into DoH. I have to write my own `CreateConnection` (that only does DoH, because that's all I care to change). And because I'm writing my own connection layer, I'm going to use `hyper` because that's what most of the community would use.

Anyway, I'd rather hold off 0.23 until we can sort out this bit of generics confusion. Let me see if I can PR in a draft that will make things a bit more clear from my end.

**Comment by djc** [comment]
> Maybe it's just confusion from naming, then. I don't know if @djc has some context on why it's designed this way, maybe that would help me understand.

As detailed in earlier review comments in this PR, I never quite got to wrap my head around all the things that this PR introduced (and disagreed with some of the naming choices), but in the end @bluejekyll decided to merge it.

> I see that there's a desire to remain compatible across both tokio and async-std, and that's where a lot of the complexity and pain comes from. There's a scratch-built http/2 client in `https_client_stream.rs` where most folks would just use `hyper`, because of course `hyper` depends on `tokio`. That part makes sense. It's rough, by necessity.

As I understand it, hyper does support async-std using a trait similar to our `Runtime` trait. If that's the case, it might make sense to have a PR that just replaces the custom http/2 client with hyper? I don't know what @bluejekyll's position is these days, but I think we did at some point discuss just dropping async-std support as its marketshare seems to have been decreasing.

> But because it's all bespoke, I can't just tweak `GenericConnection` to inject a header into DoH. I have to write my own `CreateConnection` (that only does DoH, because that's all I care to change). And because I'm writing my own connection layer, I'm going to use `hyper` because that's what most of the community would use.
> 
> Anyway, I'd rather hold off 0.23 until we can sort out this bit of generics confusion. Let me see if I can PR in a draft that will make things a bit more clear from my end.

Would be happy to see a PR that tries to tame the complexity here.

@jeff-hiner IIRC your use case is that you have a custom runtime, right, which is not Tokio or async-std?

**Comment by jeff-hiner** [comment]
@djc My use case is all under tokio, but I need to inject a custom auth header into the HTTP headers used for DoH.

The auth token rotates periodically so it's implemented not as a static configuration but as an `Fn -> impl Future`. Otherwise we'd just PR in something into the `NameServerConfig` to use optional extra static headers.

**Comment by XOR-op** [comment]
@jeff-hiner I'm sorry to hear you have to hold off 0.23 since I didn't want to drop backward compatibility for your case in this PR (just need a minor fix). But another PR about adding extra parameters to `NameServerConfig` sounds appealing. Now the `bind_addr` parameter is limited and useless, and I want to see something more generic. Would be happy to see a issue or PR to discuss this.

**Comment by jeff-hiner** [comment]
`NameServerConfig` has its own set of issues. `enum Protocol` should have  `tls_dns_name` should be a sub-field of `Protocol::Tls`, `Protocol::Https`, and `Protocol::Quic`. I could use that opportunity to add the header injection I need in there. But changing that will absolutely break existing serialization. @bluejekyll Is that acceptable?

---

I tried taking a pass at reworking the types, but everything is very tightly coupled. The issues I've encountered so far:

- `AbstractNameServer` isn't actually abstract, it's a concrete implementation with assumptions about needing to hold a runtime, the connection provider being dependent upon runtimes, explicitly passing runtime handles into fns, etc. Why is it named "Abstract"?
- `GenericConnection` is both a factory (`impl CreateConnection`) and the handle the factory produces (`impl DnsHandle`). This is confusing. Should this be `GenericConnector` (a factory) and `GenericConnection` (a handle used to send)?
- The factory end of GenericConnection needs to be generic over `<P: RuntimeProvider>` in order to be able to use UDP and TCP sockets. Instead...
- The coupling is baked into `fn new_connection<P>` assuming every `CreateConnection` impl doesn't take `&self` but instead a ref to a RuntimeProvider, which then returns a future to an owned `Self` which `impl GenericConnection`. Is `new_connection` supposed to return a handle or a factory? If the former, why does it take a RuntimeProvider instead of a factory as `&self`? If the latter, why is it returning a `CreateConnection` and not a `DnsHandle`?

**Comment by XOR-op** [comment]
> I tried taking a pass at reworking the types, but everything is very tightly coupled. The issues I've encountered so far:
> 
> * `AbstractNameServer` isn't actually abstract, it's a concrete implementation with assumptions about needing to hold a runtime, the connection provider being dependent upon runtimes, explicitly passing runtime handles into fns, etc. Why is it named "Abstract"?

Since the original comment of `NameServer` said it's used for mocking, I assumed few people would use it with this warning. However, if people really use it, we should rename it as well as the comment.

> * `GenericConnection` is both a factory (`impl CreateConnection`) and the handle the factory produces (`impl DnsHandle`). This is confusing. Should this be `GenericConnector` (a factory) and `GenericConnection` (a handle used to send)?

Sounds like a good idea. I could give it a try if I have time.

> * The factory end of GenericConnection needs to be generic over `<P: RuntimeProvider>` in order to be able to use UDP and TCP sockets. Instead...

Instead what? Can you explain it?

> * The coupling is baked into `fn new_connection<P>` assuming every `CreateConnection` impl doesn't take `&self` but instead a ref to a RuntimeProvider, which then returns a future to an owned `Self` which `impl GenericConnection`. Is `new_connection` supposed to return a handle or a factory? If the former, why does it take a RuntimeProvider instead of a factory as `&self`? If the latter, why is it returning a `CreateConnection` and not a `DnsHandle`?

I think it's the former. Do you think splitting for example `GenericConnector` and `GenericConnection` could solve this problem?

**Comment by djc** [comment]
Are the `Generic` prefixes in those names still meaningful at that point? Doesn't feel like it.

**Comment by XOR-op** [comment]
@jeff-hiner I have opened a PR for your case in #1938. It will be helpful if you could bring some comments, so that I could revise it or ask for a maintainer's review as the next step.

**Pull request merged**
